lib/ui/painting/image_decoder_impeller.cc

```diff
@@ -381,6 +381,7 @@ ImageDecoderImpeller::UnsafeUploadTextureToPrivate(
     FML_DLOG(ERROR) << decode_error;
     return std::make_pair(nullptr, decode_error);
   }
+  command_buffer->WaitUntilScheduled();
```

Comment:
Is a glFlush required here?  Can you add a comment explaining why it's required.  We aren't doing this elsewhere and it is going to add a performance hit to the decoder on opengles.

Reply:
`WaitUntilScheduled` (which wraps `glFlush`) is required here.  Without this, there is no assurance that the GPU will launch execution of the IO thread command buffer before the raster thread tries to access the texture.

Added a comment.

Reply:
> there is no assurance that the GPU will launch execution of the IO thread command buffer before the raster thread tries to access the texture.

If that's the case, are we sure we don't need a `glFinish` then?  The `glFlush` is just making sure the commands are submitted to the GPU, but `glFinish` would make it synchronize on them actually completing.  With a `glFlush` couldn't the raster thread potentially be using the texture the data is finished being uploaded?  Or maybe the driver is synchronizing on it if it isn't finished yet?

Reply:
`glFinish` is expensive - I can see a major increase in latency when scrolling quickly through the Wonderous home screen pages on a Pixel 3.  It takes several seconds for the app to catch up and render all pending images if each image decode does a `glFinish`.

However, IIUC the spec does not guarantee that a `glFlush` on the IO thread will ensure that the texture is safely usable when the raster thread consumes it on all GLES implementations.

Renamed `CommandBuffer::WaitUntilScheduled` to `CommandBuffer::WaitUntilCompleted` and changed it to call `glFinish`.

We should look into having GLES textures use `glFenceSync`/`glWaitSync` (https://github.com/flutter/flutter/issues/158963)

Reply:
Notes from the discord channel:
- https://www.khronos.org/webgl/wiki/SharedResouces says glFinish is required
- skia source uses glFinish on opengles2 and glFenceSync on opengles3